rfcs/0000-slsa-attestation.md

```diff
@@ -0,0 +1,528 @@
+
+### Documentation needed
+
+A core part of SLSA is proper documentation detailing the expected provenance and the security levels a build system
+offers. kpack's SLSA documentation must be in `docs/slsa.md` and contain a breakdown of all the fields in the provenance
+similar to the [provenance section](#provenance-schema).
+
+Example of documentation for [GitHub Actions][gha-doc] and [Google Cloud Build][gcb-doc].
+
+[gha-doc]: https://github.com/slsa-framework/github-actions-buildtypes/tree/main/workflow/v1
+[gcb-doc]: https://github.com/slsa-framework/gcb-buildtypes/tree/main/triggered-build/v1
+
+## Complexity
+
+One major implementation detail is if the attestation generation and/or signing should occur in the build pod or in the
+controller. If it's part of the build pod, it would occur in the completion step after the build metadata is read. If
+it's part of the controller, it would occur after the build pod has completed.
+
+
+### Attestation in controller
+Pros:
+- Can configure a cluster-wide signing secret
+- Easier access to `internalParameters` (i.e. passing all the internal parameters as flags to the completion step is
+  ugly)
+
+Cons:
+- Writing to registry will block reconcilation of other builds (this could be alleviated by introducing a dedicated
+  signing/attesting controller)
+- Failing to sign or push attestation can only be surfaced on the build status, not the build logs
+
+
+#### Attestation in build pod
+Pros:
+- No additional scaling problems
+- Easier access to SBoM if we expand on this in the future
+- May tap into future improvements by the CNB lifecycle
+- Any errors can be written to the build log, which is the first place users will look
+
+Cons:
+- Much harder to support a cluster-wider signing secret (since the private key might need to be shared across
+  namespaces).
+
+## Prior Art
+
+- Docker: https://docs.docker.com/build/attestations/slsa-definitions/
+- Tekton: https://tekton.dev/docs/concepts/supply-chain-security/#how-does-tekton-chains-work
+- GitHub Actions: https://github.blog/2022-04-07-slsa-3-compliance-with-github-actions/
+
+## Alternatives
+
+There's really no alternative, SLSA is designed as a way for the build platform to attest to the integrity of the build
+process and the system itself.
+
+## Risks
+
+- SLSA is relatively early in its life, while it's part of the Open Source Security Foundation (OpenSSF), industry
+  adoption is still getting started.
+
+## Additional considerations and open questions
+- Should the signing key secrets require an additional annotation instead of just matching on the type/field? This would
+  bring more inline with the [Cluste wide cosign RFC][cluster-wide-cosign-rfc]
+- Currently there's no tools on the market that can actually verify generic key-based attestations.
+  [slsa-verifier][slsa-verifier] uses keyless signing and ony supports GitHub Actions and Google Cloud Build,
+  `cosign` does key based verification, but only with its own key type.
```

Comment:
Where did we land here? How do we expect users of kpack to validate these attestations?